### PR TITLE
Collective: Benchmark with greater `MaxProposals`

### PR DESCRIPTION
**Pull request by Szegoo**
Closes: #12433

### CODE REVIEW

File: frame/system/src/lib.rs [hunk 1]
```diff
@@ -398,6 +398,15 @@ pub mod pallet {
 			Ok(().into())
 		}
 
+		#[pallet::weight(T::SystemWeightInfo::remark(_remark.len() as u32))]
+		pub fn remark_vec_u32(
+			origin: OriginFor<T>,
+			_remark: Vec<u32>,
+		) -> DispatchResultWithPostInfo {
```
Comment (Szegoo) [inline frame/system/src/lib.rs:405]: I believe that there may be a better solution to this, I will think about this, some ideas would be useful :)

File: frame/collective/src/benchmarking.rs [hunk 2]
```diff
@@ -108,7 +108,7 @@ benchmarks_instance_pallet! {
 		let b in 1 .. MAX_BYTES;
 		let m in 1 .. T::MaxMembers::get();
 
-		let bytes_in_storage = b + size_of::<u32>() as u32;
+		let bytes_in_storage = 4 * (b + size_of::<u32>() as u32);
```
Comment (Szegoo) [inline frame/collective/src/benchmarking.rs:111]: We multiply by 4 since we went from `u8` to `u32`.

**Review by ggwpez — CHANGES_REQUESTED** [review]
You can just create an `id_to_remark_data` function which converts a `u32` to a vector of data. No need to add a new extrinsic.

File: frame/collective/src/benchmarking.rs [hunk 3]
```diff
@@ -34,6 +34,10 @@ fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::
 	frame_system::Pallet::<T>::assert_last_event(generic_event.into());
 }
 
+fn id_to_remark_data<T: Config<I>, I: 'static>(id: u32, length: usize) -> SystemCall<T> {
+	SystemCall::<T>::remark { remark: vec![id.try_into().unwrap(); length] }
+}
```
Comment (Szegoo) [inline frame/collective/src/benchmarking.rs:39]: @ggwpez I don't think we can use the `remark` function for the proposal here. We have to create a valid `Call` here that is going to be the proposal in the benchmark, but we cannot use the `remark` function since it requires the argument to be a `Vec<u8>`.

Comment (ggwpez) [reply frame/collective/src/benchmarking.rs:39]: `id_to_remark_data(id: u32, len: usize) -> Vec<u8>` is what I meant.  
To test that it works you can increase the `MaxProposals` in the mock runtime to `512`.

Comment (Szegoo) [reply frame/collective/src/benchmarking.rs:39]: I don't think we can convert a `Vec<u8>` into a `Call`. The code expects a `Call`:
```rs
let proposal: T::Proposal = id_to_remark_data(i, length).into();
```
`T::Proposal` is actually `RuntimeCall`, so we cannot have `id_to_remark_data` return a `Vec<u8>`

Comment (ggwpez) [reply frame/collective/src/benchmarking.rs:39]:
```rust
let proposal: T::Proposal = SystemCall::<T>::remark { remark: id_to_remark_data(i, length) }.into();
```
You can obviously think of another way as well.

Comment (Szegoo) [reply frame/collective/src/benchmarking.rs:39]: The problem here is that we cannot have two same proposals. The remark call accepts a `Vec<u8>`. In the benchmark code, we make proposals with the same call(`remark`) but use different arguments. We submit a vector with repeated values for a specified length. These values range from `0` to `T::MaxProposals::get()`. If `T::MaxProposals::get()` is greater than `255` we run into a problem, because the `remark` function accepts only `u8`s.
I currently have two ideas on how to solve this.

1. We can change the length of the vector if the `T::MaxProposals::get()` is greater than `255`. What I mean by this is the following: We generate `remark` function calls for `0..255` with the specified length, and after we reach `255` we can increase the length by one and repeat the process. This way the calls aren't going to be the same.

2. Currently we submit repeated values for a given length to the `remark` function. What we could do instead is not repeat the same value but have the call data generated like this: from `0` to `255` we will have the following vector: `[i, repeat 0s for the given length]`, and after we reach `255` we do the same thing but we have `i` at the next index. `[0, i, repeat 0s for the given length]`...

Or maybe I am missing something and overthinking this :P

File: frame/collective/src/benchmarking.rs [hunk 4]
```diff
@@ -34,6 +34,27 @@ fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::
 	frame_system::Pallet::<T>::assert_last_event(generic_event.into());
 }
 
+fn id_to_remark_data<T: Config<I>, I: 'static>(id: u32, length: usize) -> Vec<u8> {
+	let mut data: Vec<u8> = vec![];
+	let mut value = id.clone();
+	let mut value_size = 0;
+
+	while value > u8::MAX as u32 {
```
Comment (Szegoo) [inline frame/collective/src/benchmarking.rs:42]: @ggwpez I implemented something similar to the second solution from the previous comment. It is important to note that this doesn't cover every `u32` value(the provided `length` defines the maximum `MaxProposals` size that will work), but I tested it and this works with values up to around `2048` as `MaxProposals` and I believe that no greater value than this will ever be used for benchmarking since it will be really slow.

File: frame/collective/src/benchmarking.rs [hunk 5]
```diff
@@ -34,6 +34,27 @@ fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::
 	frame_system::Pallet::<T>::assert_last_event(generic_event.into());
 }
 
+fn id_to_remark_data<T: Config<I>, I: 'static>(id: u32, length: usize) -> Vec<u8> {
+	let mut data: Vec<u8> = vec![];
+	let mut value = id.clone();
+	let mut value_size = 0;
```
Comment (Szegoo) [inline frame/collective/src/benchmarking.rs:40]: Also, this isn't the best name for this variable, didn't come up with something better, but will definitely change this.

File: frame/collective/src/benchmarking.rs (hunk 4 above)
Comment (ggwpez) [reply frame/collective/src/benchmarking.rs:42]: Yea it looks a bit better than before. Will take a look tomorrow.

File: frame/collective/src/tests.rs [hunk 6]
```diff
@@ -89,7 +89,7 @@ pub type MaxMembers = ConstU32<100>;
 
 parameter_types! {
 	pub const MotionDuration: u64 = 3;
-	pub const MaxProposals: u32 = 100;
+	pub const MaxProposals: u32 = 512;
```
Comment (Szegoo) [inline frame/collective/src/tests.rs:92]: Will revert this, only wanted to see if the benchmarks work.

File: frame/collective/src/benchmarking.rs [hunk 7]
```diff
@@ -34,6 +34,26 @@ fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::
 	frame_system::Pallet::<T>::assert_last_event(generic_event.into());
 }
 
+fn id_to_remark_data<T: Config<I>, I: 'static>(id: u32, length: usize) -> Vec<u8> {
+	// this variable contains the how many `u8::MAX` can we substract from id.
+	let mut u8_max_count = 0;
```
Comment (Szegoo) [inline frame/collective/src/benchmarking.rs:39]: Tbh I don't really like this variable name, would appreciate some ideas :)

File: frame/collective/src/benchmarking.rs [hunk 8]
```diff
@@ -34,6 +34,27 @@ fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::
 	frame_system::Pallet::<T>::assert_last_event(generic_event.into());
 }
 
+fn id_to_remark_data<T: Config<I>, I: 'static>(id: u32, length: usize) -> Vec<u8> {
+	// this variable is equal to the number of times we have to substract `u8` from id so that we
+	// get a u8.
+	let mut until_u8 = 0;
+	while id - (u8::MAX * until_u8) as u32 > u8::MAX as u32 {
```
Comment (ggwpez) [inline frame/collective/src/benchmarking.rs:41]: Sorry but what the fuck is this?

File: frame/collective/src/benchmarking.rs [hunk 9]
```diff
@@ -34,6 +34,24 @@ fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::
...
 
+fn id_to_remark_data<T: Config<I>, I: 'static>(id: u32, length: usize) -> Vec<u8> {
+	let mut remark_data: Vec<u8> = vec![];
+	let mut value = id;
+	let mut counter = 0;
+	while value > u8::MAX as u32 {
+		value -= u8::MAX as u32;
+		remark_data.push(u8::MAX);
+		counter += 1;
+	}
```
Comment (Szegoo) [inline frame/collective/src/benchmarking.rs:45]: @ggwpez Sorry for the previous solution, it was a mess now that I look back at it. Hopefully, this is a better solution, if not I will happily close this MR and learn from someone who manages to solve this in a better way.
The idea is I believe really simple here. We are creating the remark data like the following:
1. If `id < 255` we simply push the `id` to the `remark_data` and fill out the rest of the vector with zeros.
2. If not we subtract id with `255` until we get a value that is less than `255`. Until then we are pushing `255` to the `remark_data`. When id becomes less than `255` we push it to the `remark_data` and fill the rest with zeros.

For example(in this example the length is 5): id = 16 => `remark_data = [16, 0, 0, 0, 0]`
Another example(in this example the length is 5): id = 264 => `remark_data = [255, 9, 0, 0, 0]`

Comment (ggwpez) [reply frame/collective/src/benchmarking.rs:45]: If you undo these changes where you multiply with 4: `let bytes_in_storage = 4 * (b + size_of::<u32>() as u32);` then you can just do: `id.to_le_bytes().cycle().take().collect()` in this function and be done with it.

Comment (Szegoo) [reply frame/collective/src/benchmarking.rs]: Clever solution.

File: frame/collective/src/benchmarking.rs [hunk 10]
```diff
@@ -34,6 +34,10 @@ fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::
 	frame_system::Pallet::<T>::assert_last_event(generic_event.into());
 }
 
+fn id_to_remark_data<T: Config<I>, I: 'static>(id: u32, length: usize) -> Vec<u8> {
```
Comment (ggwpez) [inline frame/collective/src/benchmarking.rs:37]:
```suggestion
fn id_to_remark_data(id: u32, length: usize) -> Vec<u8> {
```

File: frame/collective/src/tests.rs (hunk 6 above)
Comment (ggwpez) [inline frame/collective/src/tests.rs:92]: Please add a small test that fails if `MaxProposals` is larger than 2^16, since we know that the benchmarks would not work with that.

Comment (ggwpez) [reply frame/collective/src/tests.rs:92]: If we revert it then a future developer could accidentally break it.  
`257` could be a middle ground where it would find trivial collisions but still be reasonably fast without `--release` flag.

**Review by ggwpez — APPROVED** [review]
Good now.

File: frame/collective/src/tests.rs (hunk 6 above)
Comment (Szegoo) [reply frame/collective/src/tests.rs]: It shouldn't fail at 2^16, 2^32 should be the value when it fails.

Comment (Szegoo) [reply frame/collective/src/tests.rs]: We could even restrict `MaxPropsoal` to `u16` here, or increase the length everywhere so that we can use `u32`(Tbh I don't see the point of that since that would really slow down the benchmarks)

Comment (ggwpez) [reply frame/collective/src/tests.rs]: We normally use `u32` as convention, even though the benchmark will most likely never work with 2³².  
Using 4 byte would increase the base weight of the benchmarking result, so its better to keep the lower bounds low.

**Comment by Szegoo** [comment]
This PR is waiting for review.

File: frame/collective/src/benchmarking.rs [hunk 11]
```diff
@@ -529,7 +527,7 @@ benchmarks_instance_pallet! {
 	}
 
 	close_approved {
-		let b in 1 .. MAX_BYTES;
+		let b in 2 .. MAX_BYTES;
```
Comment (bkchr) [inline frame/collective/src/benchmarking.rs:530]: Why these changes?

**Review by bkchr — APPROVED** [review]
Looks good, just not sure why `MAX_BYTES` changed?

File: frame/collective/src/benchmarking.rs (hunk 11 above)
Comment (Szegoo) [reply frame/collective/src/benchmarking.rs:530]: The goal of this PR was to be able to set the `MaxProposal` to a value higher than `255`. In the benchmarks, we use the `remark` function to represent a proposal. Because we cannot have two identical proposals what we did until now is make the remark data be a variable number repeated a number of times(`b` times). This comes with a limit since the remark data is a `Vec<u8>` so it cannot accept a number greater than `255`. To fix this we use little-endian to encode these greater numbers. So to represent a number greater than `255` we need a vector that is longer than 1. That is why we updated the range of  `b` so that it starts from 2.

Comment (bkchr) [reply frame/collective/src/benchmarking.rs:530]: Then it should be changed to `4`, because `MaxMembers` is an `u32` which would lead to duplicate proposals when using more than 65k max members. You should also leave some comment on why we start at 4.

Comment (Szegoo) [reply frame/collective/src/benchmarking.rs:530]: https://github.com/paritytech/substrate/pull/12454#discussion_r994730425

Comment (bkchr) [reply frame/collective/src/benchmarking.rs:530]: Hmm okay :see_no_evil:

Comment (bkchr) [reply frame/collective/src/benchmarking.rs:530]: I'm not happy with this, but okay :P

**Comment by bkchr** [comment]
bot merge
